### PR TITLE
📝 Document the Usage of Templates and add Rationale of QDMI

### PR DESCRIPTION
**Pull request by ystade**
This PR adds a new page to the documentation that explains how to use the template provided together with QDMI. It adds another page explaining the motivation behind some design decisions within QDMI.

### Remarks

This PR additionally includes the following changes:
- Updates the commit hash in the template
- Updates the logo, esp, makes it adapting to dark and bright mode
- Move C++ examples in examples to the front.
- Add the logo to the README.md.
- Makes the QDMI diagram adapt to the dark mode

### CODE REVIEW

**Comment by ystade** [comment]
I was wondering whether we actually want to fix the commit hash in the template or whether we just fetch the latest version on the default branch. What do you think, @burgholzer ?

**Comment by burgholzer** [comment]
> I was wondering whether we actually want to fix the commit hash in the template or whether we just fetch the latest version on the default branch. What do you think, @burgholzer ?

Just for the sake of not having to regularly update the template, I think it makes most sense to use the default branch and add a comment that, once instantiated, people should be adjusting this to either a commit hash or a version pin to guarantee stability.

**Comment by ystade** [comment]
Two questions:
- Is there a way to check out the generated GitHub page before merging except locally?
- Where should the logo be placed in the README.md??

**Comment by burgholzer** [comment]
> Two questions:
> 
> * Is there a way to check out the generated GitHub page before merging except locally?

Guess locally is the only real option here. But given how the GitHub page simply uses the output from doxygen and hosts that statically, that should be good enough, right?

> * Where should the logo be placed in the README.md??

Maybe similarly to how we place it for the MQT repositories, i.e., right on top of the top-level header.
The current image could then be moved below the header.

**Comment by ystade** [comment]
Then I guess, it is ready from my side.

File: docs/_static/qdmi.svg
Comment (burgholzer) [inline docs/_static/qdmi.svg:1]: Should we replace the term "Platform" with "Device" in these Graphics?
More generally: Should we update this with the latest version from the MQSF presentation? See screenshot
image

File: docs/index.md
Comment (burgholzer) [inline docs/index.md:1]: Should we populate the end of this page with a couple of pointers to the various documents that can be found in the documentation? I think that could help people find stuff.

File: docs/rationale.md
Comment (burgholzer) [inline docs/rationale.md:1]: Should this file contain a couple more crossreferences to the API docs?

**Review by burgholzer — APPROVED** [review]
Hey @ystade 👋🏼 

Thanks for all your work in this!
The documentation is really looking great now!

I edited a couple of minor things and left a hand full of inline comments which probably should be addressed. Otherwise, this is good to go from my side 🚀

File: docs/index.md
Comment (ystade) [reply docs/index.md:1]: Take a look at  [4210735](https://github.com/Munich-Quantum-Software-Stack/QDMI/commit/4210735b3773f057e692b431b7f788630acec9fd).

File: docs/rationale.md
Comment (ystade) [reply docs/rationale.md:1]: Take a look at [c2188bf](https://github.com/Munich-Quantum-Software-Stack/QDMI/commit/c2188bffed1362346ecbdcf5e487293e49697bb8). ;)

File: docs/_static/qdmi.svg
Comment (ystade) [reply docs/_static/qdmi.svg:1]: I would say, either we wait until the flyer is done and we take pics from that or we merge that as is and update the pics later on.

Comment (burgholzer) [reply docs/_static/qdmi.svg:1]: In that case I am all in favor of merging this as is and updating later!
Let's keep the PRs moving 😌

File: docs/rationale.md
Comment (burgholzer) [reply docs/rationale.md:1]: LGTM. I think there are a couple more references to particular "types" where a cross reference could be used, but we can add these later!

File: docs/index.md [hunk 1]
```diff
@@ -6,8 +6,28 @@
...
 
 \snippet{doc} README.md DOXYGEN MAIN
+
+### How to Use this Documentation?
+
+This documentation tries to provide helpful information to get you started with QDMI.
+
+#### General Information
+
+The [FQA](faq.md) page gives an overview over frequently asked questions. More details about the
```
Comment (burgholzer) [inline docs/index.md]:
```suggestion
The [FAQ](faq.md) page gives an overview over frequently asked questions. More details about the
```

File: docs/index.md [hunk 2]
```diff
@@ -6,8 +6,28 @@
...
+
+#### Hands-On
+
+The [Examples](examples.md) page provides a collection of examples to see QDMI in action. When you
+want to start your own implementation of QDMI, the [Templates](templates.md) page is a good starting
+point.
+
+#### Nitty-Gritty Details
+
+When you are interested in the details of the semantics of function, the [API Reference](files.html)
```
Comment (burgholzer) [inline docs/index.md]:
```suggestion
When you are interested in the details of the semantics of functions and the whole API, the [Reference Documentation](files.html)
```

File: docs/index.md
Comment (burgholzer) [reply docs/index.md:1]: Just two small comments on wording. Otherwise this LGTM 👍🏻

File: docs/rationale.md
Comment (ystade) [reply docs/rationale.md:1]: I have added further refs in [3b93f35](https://github.com/Munich-Quantum-Software-Stack/QDMI/pull/90/commits/3b93f35f9dbe7b70270dcc3b00c8c98f614eca00)

**Pull request merged**